### PR TITLE
Added Method yii\web\Request::getDomain()

### PR DESCRIPTION
**Pull request by ivanovyordan**
| Q | A |
| --- | --- |
| Is bugfix? | no |
| New feature? | yes |
| Breaks BC? | no |
| Tests pass? | yes |
| Fixed issues | #11309 |

Also modified `yii\web\Request::getHostInfo()` to use the new method.
P.S. Not sure what kind of unit test I need to write to cover the new methods.

### CODE REVIEW

**Comment by rob006** [comment]
```php
Yii::$app->getRequest()->hostInfo = 'http://example.com';
echo Yii::$app->getRequest()->domain;
```

You will expect that `Request::getDomain()` will return `example.com`, but you will get completely different result.

**Comment by tunecino** [comment]
please check those:
http://stackoverflow.com/questions/288810/get-the-subdomain-from-a-url
http://stackoverflow.com/questions/399932/can-i-improve-this-regex-check-for-valid-domain-names
https://github.com/yiisoft/yii2/issues/11309#issuecomment-207453311

**Comment by rob006** [comment]
@tunecino This is irrelevant in this context. If `Request::hostInfo` is `http://myapp.example.com` I expect that `Request::getDomain()` will return `myapp.example.com` instead of `example.com`.

**Comment by tunecino** [comment]
mmm I see. you mean like what `parse_url($url)['host']` would return. I guess I did misunderstood this from the beginning.

File: framework/web/Request.php [hunk 1]
```diff
@@ -520,6 +522,48 @@ public function getQueryParam($name, $defaultValue = null)
...
 
+    private $_domain;
+
+    /**
+     * Returns host part of the current request URL.
+     * By default this is determined based on the user request information.
+     * You may explicitly specify it by setting the [[setDomain()|domain]] property.
+     * @return string hostname part of the request URL (e.g. `yiiframework.com`),
+     * null if can't be obtained from `$_SERVER` and wasn't set.
+     * @see setDomain()
```
Comment (samdark) [inline framework/web/Request.php]: Need `@since` tag.

File: framework/web/Request.php [hunk 2]
```diff
@@ -520,6 +522,48 @@ public function getQueryParam($name, $defaultValue = null)
...
+     */
+    public function getDomain()
+    {
+        if ($this->_domain === null) {
+            if (isset($_SERVER['HTTP_HOST'])) {
+                $this->_domain = $_SERVER['HTTP_HOST'];
+                if(strpos($this->_domain, ':') !== false) {
```
Comment (samdark) [inline framework/web/Request.php]: Formatting. Need space before `(`.

File: framework/web/Request.php [hunk 3]
```diff
@@ -520,6 +522,48 @@ public function getQueryParam($name, $defaultValue = null)
...
+        }
+
+        return $this->_domain;
+    }
+
+    /**
+     * Sets the host part of the application URL.
+     * This setter is provided in case the hostname cannot be determined
+     * on certain Web servers.
+     * @param string $value the host part of the application URL.
```
Comment (samdark) [inline framework/web/Request.php]: Need `@since` tag.

File: framework/web/Request.php [hunk 4]
```diff
@@ -535,15 +579,11 @@ public function getHostInfo()
     {
         if ($this->_hostInfo === null) {
             $secure = $this->getIsSecureConnection();
+            $domain = $this->getDomain();
             $http = $secure ? 'https' : 'http';
-            if (isset($_SERVER['HTTP_HOST'])) {
-                $this->_hostInfo = $http . '://' . $_SERVER['HTTP_HOST'];
-            } elseif (isset($_SERVER['SERVER_NAME'])) {
-                $this->_hostInfo = $http . '://' . $_SERVER['SERVER_NAME'];
+            if($domain !== null) {
```
Comment (samdark) [inline framework/web/Request.php]: Need space after `if`

File: framework/web/Request.php [hunk 5]
```diff
@@ -535,15 +579,11 @@ public function getHostInfo()
...
                 $port = $secure ? $this->getSecurePort() : $this->getPort();
-                if (($port !== 80 && !$secure) || ($port !== 443 && $secure)) {
-                    $this->_hostInfo .= ':' . $port;
-                }
+                $this->_hostInfo = $http . '://' . $domain . ':' . $port;
```
Comment (samdark) [inline framework/web/Request.php]: I think it changes behavior i.e. `:80` and `:433` will be added while previously it was not.

**Review by samdark — CHANGES_REQUESTED** [review]

**Comment by rob006** [comment]
Alternative: #12686

**Comment by samdark** [comment]
@ivanovyordan what do you think about @rob006 solution?

**Comment by samdark** [comment]
https://github.com/yiisoft/yii2/pull/12686 merged. Thanks for contributing a possible solution.

**Pull request closed without merging**